bundles/org.openhab.binding.upnpcontrol/src/main/java/org/openhab/binding/upnpcontrol/internal/handler/UpnpServerHandler.java

```diff
@@ -0,0 +1,471 @@
+import org.openhab.binding.upnpcontrol.internal.UpnpXMLParser;
+import org.openhab.binding.upnpcontrol.internal.config.UpnpControlServerConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link UpnpServerHandler} is responsible for handling commands sent to the UPnP Server.
+ *
+ * @author Mark Herwege - Initial contribution
+ * @author Karel Goderis - Based on UPnP logic in Sonos binding
+ */
+@NonNullByDefault
+public class UpnpServerHandler extends UpnpHandler {
+
+    private static final String DIRECTORY_ROOT = "0";
+    private static final String UP = "..";
+
+    private final Logger logger = LoggerFactory.getLogger(UpnpServerHandler.class);
+
+    private ConcurrentMap<String, UpnpRendererHandler> upnpRenderers;
+    private volatile @Nullable UpnpRendererHandler currentRendererHandler;
+    private volatile List<StateOption> rendererStateOptionList = new ArrayList<>();
+
+    private @NonNullByDefault({}) ChannelUID rendererChannelUID;
+    private @NonNullByDefault({}) ChannelUID currentTitleChannelUID;
+
+    private volatile int numberReturned;
+    private volatile int totalMatches;
+
+    private volatile UpnpEntry currentEntry = new UpnpEntry(DIRECTORY_ROOT, DIRECTORY_ROOT, DIRECTORY_ROOT,
+            "object.container");
+    private volatile List<UpnpEntry> entries = new ArrayList<>(); // current entry list in selection
+    private volatile Map<String, UpnpEntry> parentMap = new HashMap<>(); // store parents in hierarchy separately to be
+                                                                         // able to move up in directory structure
+
+    private List<String> source = new ArrayList<>();
+
+    private UpnpDynamicStateDescriptionProvider upnpStateDescriptionProvider;
+    private UpnpDynamicCommandDescriptionProvider upnpCommandDescriptionProvider;
+
+    protected UpnpControlServerConfiguration config = getConfigAs(UpnpControlServerConfiguration.class);
+
+    public UpnpServerHandler(Thing thing, UpnpIOService upnpIOService,
+            ConcurrentMap<String, UpnpRendererHandler> upnpRenderers,
+            UpnpDynamicStateDescriptionProvider upnpStateDescriptionProvider,
+            UpnpDynamicCommandDescriptionProvider upnpCommandDescriptionProvider) {
+        super(thing, upnpIOService);
+        this.upnpRenderers = upnpRenderers;
+        this.upnpStateDescriptionProvider = upnpStateDescriptionProvider;
+        this.upnpCommandDescriptionProvider = upnpCommandDescriptionProvider;
+
+        // put root as highest level in parent map
+        parentMap.put(currentEntry.getId(), currentEntry);
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
+
+        logger.debug("Initializing handler for media server device {}", thing.getLabel());
+
+        rendererChannelUID = thing.getChannel(UPNPRENDERER).getUID();
+        currentTitleChannelUID = thing.getChannel(BROWSE).getUID();
```

Comment:
`getChannel()` can return null. Same for below.

Reply:
@fwolter These channels are statically defined in the thing-types.xml file. According to the method declaration, they would only return null if no channel with the ID exists. Can there be a situation where the statically defined channel does not exist yet at this point in the code? Wouldn't that be an error in OH core?

Reply:
It will return null if one of these Channels would be renamed or removed in further versions and the user wants to continue using his Things.

Reply:
OK. I changed it. But wouldn't the same argument apply when calling updateState(String, String) where the first argument is the channel ID string? I don't do a check in that case either.

Reply:
Calling `updateState(String, String)` on a non-existent ChannelUID is safe. See org.eclipse.smarthome.core.thing.internal.CommunicationManager:483.